File: 수학_소수/프로그래머스Lv0-소인수분해.py
from collections import defaultdict
def make_primes(n:int)->list:
    primes = [True] * (n + 1)
    primes[0:2] = [False, False]
    for num in range(2, int(n * 0.5) + 1):
        if primes[num]:
            for i in range(num ** 2, n + 1, num):
                primes[i] = False
    return primes

def make_prime_factorization(n):
    primes = make_primes(n)
    prime_numbers = [x for x in range(2, n + 1) if primes[x]]
    result = defaultdict(int)
    for number in prime_numbers:
        while n % number == 0 and n > 1:
            result[number] += 1
            n //= number
        if n == 1:
            break
    return result

def solution(n):
    prime_factorization = make_prime_factorization(n)
    return sorted(prime_factorization.keys())
"""
정확성  테스트
테스트 1 〉	통과 (1.93ms, 10.1MB)
테스트 2 〉	통과 (0.01ms, 10.1MB)
테스트 3 〉	통과 (0.02ms, 10.3MB)
테스트 4 〉	통과 (0.03ms, 10.3MB)
테스트 5 〉	통과 (0.84ms, 10.3MB)
테스트 6 〉	통과 (0.18ms, 10.1MB)
테스트 7 〉	통과 (0.02ms, 10.2MB)
테스트 8 〉	통과 (0.12ms, 10.3MB)
테스트 9 〉	통과 (2.80ms, 10.3MB)
테스트 10 〉	통과 (0.65ms, 10MB)
테스트 11 〉	통과 (0.99ms, 10.1MB)
테스트 12 〉	통과 (0.83ms, 10.1MB)
테스트 13 〉	통과 (1.24ms, 10.2MB)
테스트 14 〉	통과 (1.54ms, 10.1MB)
테스트 15 〉	통과 (1.24ms, 10.2MB)
테스트 16 〉	통과 (2.00ms, 10.3MB)
테스트 17 〉	통과 (0.01ms, 10.4MB)
테스트 18 〉	통과 (0.01ms, 10.2MB)
테스트 19 〉	통과 (0.01ms, 10.2MB)
테스트 20 〉	통과 (0.01ms, 10.1MB)
테스트 21 〉	통과 (0.01ms, 10.3MB)
테스트 22 〉	통과 (0.01ms, 10.3MB)
테스트 23 〉	통과 (0.03ms, 10.2MB)
테스트 24 〉	통과 (0.64ms, 10.2MB)
"""

File: 수학_소수/프로그래머스Lv0-소인수분해.js
function makePrimes(n) {
  const primes = Array.from({ length: n + 1 }, (v) => true)
  primes.splice(0, 2, false, false)
  for (let num = 2; num < Math.floor(Math.sqrt(n)) + 1; num++) {
    if (primes[num]) {
      for (let i = num * num; i < n + 1; i += num) {
        primes[i] = false
      }
    }
  }
  return primes
}

function makePrimeFactorization(n) {
  const primes = makePrimes(n)
  const primeNumbers = Array.from({ length: n + 1 }, (v, i) => i).filter((x) => primes[x])
  const result = {}
  for (const number of primeNumbers) {
    if (n % number === 0) {
      result[number] = 0
    }
    while (n % number === 0 && n > 1) {
      result[number]++
      n /= number
    }
    if (n === 1) break
  }
  return result
}

function solution(n) {
  const primeFactorization = makePrimeFactorization(n)
  return Object.keys(primeFactorization)
    .map((x) => Number(x)) // 자바스크립트 key는 문자열이다.
    .sort((a, b) => a - b)
}
// 정확성  테스트
// 테스트 1 〉	통과 (2.44ms, 33.7MB)
// 테스트 2 〉	통과 (0.15ms, 33.4MB)
// 테스트 3 〉	통과 (0.29ms, 33.4MB)
// 테스트 4 〉	통과 (0.28ms, 33.4MB)
// 테스트 5 〉	통과 (1.53ms, 33.7MB)
// 테스트 6 〉	통과 (0.56ms, 33.6MB)
// 테스트 7 〉	통과 (0.28ms, 33.4MB)
// 테스트 8 〉	통과 (0.43ms, 33.6MB)
// 테스트 9 〉	통과 (2.46ms, 33.6MB)
// 테스트 10 〉	통과 (0.78ms, 33.5MB)
// 테스트 11 〉	통과 (2.11ms, 33.6MB)
// 테스트 12 〉	통과 (1.64ms, 33.6MB)
// 테스트 13 〉	통과 (2.05ms, 33.8MB)
// 테스트 14 〉	통과 (2.36ms, 33.7MB)
// 테스트 15 〉	통과 (2.01ms, 33.8MB)
// 테스트 16 〉	통과 (1.91ms, 33.7MB)
// 테스트 17 〉	통과 (0.17ms, 33.5MB)
// 테스트 18 〉	통과 (0.16ms, 33.5MB)
// 테스트 19 〉	통과 (0.16ms, 33.6MB)
// 테스트 20 〉	통과 (0.35ms, 33.5MB)
// 테스트 21 〉	통과 (0.16ms, 33.5MB)
// 테스트 22 〉	통과 (0.28ms, 33.4MB)
// 테스트 23 〉	통과 (0.29ms, 33.5MB)
// 테스트 24 〉	통과 (1.50ms, 33.5MB)
